pkg/cloud/volume_limits.go

```diff
@@ -0,0 +1,282 @@
+package cloud
+
+import (
+	"regexp"
+	"strings"
+)
+
+/// https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instance-types.html#ec2-nitro-instances
+const (
+	highMemoryMetalInstancesMaxVolumes   = 19
+	highMemoryVirtualInstancesMaxVolumes = 27
+	baremetalMaxVolumes                  = 31
+	nonNitroMaxAttachments               = 39
+	nitroMaxAttachments                  = 28
+)
+
+/// It is possible to have an instance family where the virtualized instances are Nitro
+/// and metal instances are not
+var nonNitroInstances = map[string]struct{}{
+	"c6i.metal": {},
+	"g5g.metal": {},
+}
+
+/// List of nitro instance types can be found here: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instance-types.html#ec2-nitro-instances
+var nonNitroInstanceFamilies = map[string]struct{}{
+	"t2":     {},
+	"c7g":    {},
+	"c4":     {},
+	"r4":     {},
+	"x2idn":  {},
+	"x2iedn": {},
+	"x2iezn": {},
+	"x1e":    {},
+	"x1":     {},
+	"p2":     {},
+	"tr1n":   {},
+	"g4dn":   {},
+	"g3":     {},
+	"d2":     {},
+	"h1":     {},
+}
+
+func IsNitroInstanceType(it string) bool {
+	if _, ok := nonNitroInstances[it]; ok {
+		return false
+	}
+
+	strs := strings.Split(it, ".")
+
+	if len(strs) != 2 {
+		panic("cannot determine family of instance type")
+	}
+
+	family := strs[0]
+	_, ok := nonNitroInstanceFamilies[family]
+	return !ok
+}
+
+func GetMaxAttachments(nitro bool) int {
+	if nitro {
+		return nitroMaxAttachments
+	}
+	return nonNitroMaxAttachments
+}
+
+/// Some instance types have a maximum limit of EBS volumes
+/// https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instance-types.html#ec2-nitro-instances
+var maxVolumeLimits = map[string]int{
+	"d3.8xlarge":    3,
+	"d3.12xlarge":   3,
+	"inf1.xlarge":   26,
+	"inf1.2xlarge":  26,
+	"inf1.6xlarge":  23,
+	"inf1.24xlarge": 11,
+	"mac1.metal":    16,
+}
+
+func GetEBSLimitForInstanceType(it string) (int, bool) {
+	if v, ok := maxVolumeLimits[it]; ok {
+		return v, ok
+	}
+
+	highMemoryMetalRegex := `^u-[a-z0-9]+\.metal$`
+	re := regexp.MustCompile(highMemoryMetalRegex)
+
+	if ok := re.MatchString(it); ok {
+		return highMemoryMetalInstancesMaxVolumes, true
+	}
+
+	highMemoryVirtualRegex := `^u-[a-z0-9]+\.[a-z0-9]+`
+	re = regexp.MustCompile(highMemoryVirtualRegex)
+
+	if ok := re.MatchString(it); ok {
+		return highMemoryVirtualInstancesMaxVolumes, true
+	}
+
+	bareMetalRegex := `[a-z0-9]+\.metal$`
+	re = regexp.MustCompile(bareMetalRegex)
+
+	if ok := re.MatchString(it); ok {
+		return baremetalMaxVolumes, true
+	}
+
+	return 0, false
+}
+
+func GetNVMeInstanceStoreVolumesForInstanceType(it string) int {
+	if v, ok := nvmeInstanceStoreVolumes[it]; ok {
+		return v
+	}
+	return 0
+}
+
+/// https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/InstanceStorage.html#instance-store-volumes
+/// IMDS does not provide NVMe instance store data; we'll just list all instances here
+var nvmeInstanceStoreVolumes = map[string]int{
+	"c5ad.large":     1,
```

Comment:
Can we get these config from describeInstanceType API instead of hard code the value?

Reply:
Yeah, it's an ugly solution, but I didn't want to make another API call just to get that information. It's possible we can do it just once at the start, but we would also need to update the policy if we want to do that.